swarm/benches/connection_handler.rs

```diff
@@ -0,0 +1,357 @@
+    where
+        Self: Default + NetworkBehaviour,
+    {
+        let mut swarm_a = new_swarm(Self::default());
+        let mut swarm_b = new_swarm(Self::default());
+
+        let behaviour_count = swarm_a.behaviours().len();
+        let protocol_count = behaviour_count * protocols_per_behaviour;
+        let protocols = (0..protocol_count)
+            .map(|i| StreamProtocol::new(format!("/protocol/{i}").leak()))
+            .collect::<Vec<_>>()
+            .leak();
+
+        let mut protocol_chunks = protocols.chunks(protocols_per_behaviour);
+        swarm_a.for_each_beh(|b| b.protocols = protocol_chunks.next().unwrap());
+        let mut protocol_chunks = protocols.chunks(protocols_per_behaviour);
+        swarm_b.for_each_beh(|b| b.protocols = protocol_chunks.next().unwrap());
+
+        swarm_a.for_each_beh(|b| b.spam_count = spam_count);
+        swarm_b.for_each_beh(|b| b.spam_count = 0);
+
+        swarm_a.for_each_beh(|b| b.other_peer = Some(*swarm_b.local_peer_id()));
+        swarm_b.for_each_beh(|b| b.other_peer = Some(*swarm_a.local_peer_id()));
+
+        static mut OFFSET: usize = 0;
+        let offset = unsafe {
+            OFFSET += 1;
+            OFFSET
+        };
+
+        swarm_a
+            .listen_on(format!("/memory/{offset}").parse().unwrap())
+            .unwrap();
+        swarm_b
+            .dial(format!("/memory/{offset}").parse::<Multiaddr>().unwrap())
+            .unwrap();
+
+        let name = format!(
+            "{}::bench().poll_count({}).protocols_per_behaviour({})",
+            std::any::type_name::<Self>(),
+            spam_count,
+            protocols_per_behaviour
+        );
+
+        c.bench_function(&name, |b| {
+            swarm_a.for_each_beh(|b| b.finished = false);
+            swarm_b.for_each_beh(|b| b.finished = false);
+            b.iter(|| {
+                futures::executor::block_on(async {
+                    while swarm_a.any_beh(|b| !b.finished) || swarm_b.any_beh(|b| !b.finished) {
+                        futures::future::select(swarm_b.next(), swarm_a.next()).await;
+                    }
+                });
+            });
+        });
+    }
+}
+
+impl<T: BigBehaviour + NetworkBehaviour> BigBehaviour for libp2p_swarm::Swarm<T> {
+    fn behaviours(&mut self) -> &mut [PollerBehaviour] {
+        self.behaviour_mut().behaviours()
+    }
+}
+
+fn new_swarm<T: NetworkBehaviour>(beh: T) -> libp2p_swarm::Swarm<T> {
+    let swarm_a_keypair = libp2p_identity::Keypair::generate_ed25519();
+    libp2p_swarm::Swarm::new(
+        MemoryTransport::new()
+            .upgrade(multistream_select::Version::V1)
+            .authenticate(libp2p_plaintext::Config::new(&swarm_a_keypair))
+            .multiplex(libp2p_yamux::Config::default())
+            .boxed(),
+        beh,
+        swarm_a_keypair.public().to_peer_id(),
+        libp2p_swarm::Config::with_tokio_executor().with_idle_connection_timeout(Duration::MAX),
```

Comment:
You are running on the tokio executor meaning the actually `Connection` runs on an entirely separate task in the tokio runtime and not as part of `Swarm::next` which is what we are benchmarking. I don't think any of those numbers can be trusted at the moment.

You want to use the following ctor and initialize it with `None` to force all connections to run on the current task: https://github.com/libp2p/rust-libp2p/blob/9bb2a871939ded657d380cc2b7beb519840f3e8d/swarm/src/connection/pool.rs#L974

See https://github.com/libp2p/rust-libp2p/blob/9bb2a871939ded657d380cc2b7beb519840f3e8d/swarm/src/connection/pool.rs#L793.

Sorry about that, I should have thought of this earlier.

The above API is crate-private so we can't access it externally. You can experiment with making it public to see some different numbers. We used to have a public API for running without an executor but we removed that because it is not useful in practice.

You've already done all of the leg work to dynamically generate the benchmarks. Perhaps moving to https://nikolaivazquez.com/blog/divan/ which can be embedded anywhere in the code and thus access `pub(crate)` functions isn't too hard? Disclaimer: I've never used `divan` before and it is new so maybe it doesn't work for our case.